### PR TITLE
feat(prims-ts): add non-absorbed MLA prefill

### PR DESCRIPTION
**Pull request by PerkzZheng**
## 📌 Description

Support contiguous non-absorbed MLA prefill with separate Q/K=192 and V/O=128 in `prims_ts`, targeting BF16 and FP8. Reuse paired 128-row kernels with 128+64 QK staging, independent V/output sizing, and improved TMEM/softmax overlap. Serialize correction for single-accumulator schedules to prevent a PV overwrite race exposed by faster softmax.

## 🔍 Related Issues

None.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [ ] All tests are passing (`unittest`, etc.).

GB300: 291 context/mask accuracy cases passed, including 64 MLA cases across batches 1/2/4, fixed/ragged lengths, BF16/FP8 input/output, and dense/causal masks. Coverage includes large masked scores and CUDA-graph replay after Q/KV length changes without replanning. Memcheck and synccheck each passed four cases with zero errors; pre-commit and version-gate checks passed. Hosted GPU CI has not validated this head.

## 🔬 Experimental Track

- [ ] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #
  - [ ] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [ ] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [ ] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [ ] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [ ] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
# One target per line: a directory or a file. (A pytest ::selector is not
# supported -- the sharding runner cannot consume one.) Must be under
# tests/experimental/ and must exist. Delete these comment lines and add yours, e.g.
#
#   tests/experimental/test_my_backend.py
#   tests/experimental/my_backend/
#
# Declaring the whole tree (tests/experimental/) is allowed but means every
# experimental PR pays for every other feature's tests, in every matrix cell.
```

## Reviewer Notes

GB300, causal, batch 1, Hq/Hkv=96/1, QK/V=192/128, BF16 output. Variable-length packed plans use uniform values in runtime length arrays. Runtime tile visibility preserves hardware score maxima away from the causal boundary; boundary tiles recompute masked maxima. Latest median latencies:

| Input | Sequence | prims_ts (ms) | trtllm-gen (ms) |
|---|---:|---:|---:|
| BF16 | 8192 | 1.146 | 1.171 |
| BF16 | 16384 | 4.323 | 4.341 |
| FP8 E4M3 | 8192 | 0.827 | 0.813 |
| FP8 E4M3 | 16384 | 2.990 | 2.876 |

Unified `BatchPrefillWithRaggedKVCacheWrapper` benchmark, `--backends prims-ts trtllm-native --refcheck`, CUPTI with CUDA graphs and cold L2, 10 warmups/30 iterations, three runs. All sampled FP32 and graph-replay checks passed. Every BF16/FP8 paired run is within 5% of trtllm-gen (worst gap 4.07%). Existing D128/D256 BF16/FP8 paths showed no regression in two paired 8K runs. The stock trtllm-gen path also returns LSE. B200 and SM107 MLA performance are unmeasured.

## Summary by CodeRabbit

- **New Features**
  - Added support for non-absorbed MLA attention with separate QK and value head dimensions of 192 and 128.
  - Improved scheduling, memory handling, output sizing, and CUDA graph replay for supported MLA configurations.
  - Added applicable LDTM.STAT optimization paths on newer GPU environments.
  - Updated supported backend head-dimension combinations and validation.

- **Documentation**
  - Updated supported configurations, layouts, limitations, scheduling details, and usage examples, including SM107 validation status.

- **Tests**
  - Expanded coverage for MLA prefill configurations, output shapes, CUDA graph replay, and validation behavior.

### CODE REVIEW

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 1]
```diff
@@ -1848,23 +1855,34 @@ def _configure_single_instance_warp_layout(cfg: FmhaConfig) -> None:
...
-    if cfg.num_qkv_instances != 1:
+    if cfg.num_qkv_instances != 1 and cfg.logical_head_dim_qk != 192:
         return
     cfg.head_dim_per_stage_kv = 128
     cfg.num_head_dim_stages_k = cfg.qk_mma_tiler[2] // cfg.head_dim_per_stage_kv
     cfg.num_head_dim_stages_v = cfg.pv_mma_tiler[1] // cfg.head_dim_per_stage_kv
-    cfg.num_o_head_dim_stages = cfg.epi_tile[1] // cfg.head_dim_per_stage_kv
+    # Two BF16 Q tiles plus three K/V stages leave room for two 64-wide
+    # output staging buffers. FP8 operands/output can retain 128-wide stores.
+    cfg.head_dim_per_stage_o = (
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:1889]: how does it differ with previous implementation ?

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 2]
```diff
@@ -2450,9 +2472,23 @@ def __init__(
         v_dtype = in_dtype or cutlass.Float16
         o_dtype = out_dtype or cutlass.Float16
 
+        d_v = d if d_v is None else d_v
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:2510]: can you add comments to mention that this code block is for head dim validation.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 3]
```diff
@@ -2450,9 +2472,23 @@ def __init__(
...
+        if (d_v != d or d == 192) and (d, d_v) != (192, 128):
+            raise ValueError("asymmetric context requires QK=192 and V=128")
+        if d == 192 and (head_paired or use_paged_kv):
+            raise NotImplementedError(
+                "192/128 requires contiguous query-paired context"
+            )
+        # TMA fills the partial final Q/K slice with zeros. Global tensors
+        # retain their compact 192-element rows; only SMEM is padded.
+        padded_d = 256 if d == 192 else d
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py]: can you make it as general like so different d can pad to 64, 128, 256 separately.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py [hunk 4]
```diff
@@ -2704,6 +2740,13 @@ def __call__(
                 stride_order=stride_order,
                 swizzle=tma_qkv_swizzle,
                 l2_promotion=cuda.TensorMapL2Promotion.none,
+                # The 192-wide Q tensor has an OOB head-dimension tail.
+                # Ragged TMA's default NaN fill would poison those QK FMAs.
+                oob_fill=(
+                    cuda.TensorMapFloatOOBFill.none
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py]: we should always set it cuda.TensorMapFloatOOBFill.none

File: tests/attention/test_attention_ts_context_mla.py
Comment (PerkzZheng) [inline tests/attention/test_attention_ts_context_mla.py:1]: Move this to test_attention_ts_context.py instead.

File: tests/attention/test_attention_ts_context.py [hunk 5]
```diff
@@ -950,6 +950,10 @@ def test_attention_ts_context_uses_ldtm_stat_default_follows_gpu():
         context_module._dsl_supports_ldtm_stat()
         and torch.cuda.get_device_capability() in ((10, 3), (10, 7))
     )
+    if torch.cuda.get_device_capability() == (10, 7):
```
Comment (PerkzZheng) [inline tests/attention/test_attention_ts_context.py:2646]: ldtm reduce is also supported by 10, 3.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py [hunk 6]
```diff
@@ -948,6 +977,24 @@ def mma_schedule(
                 desc_k_base=desc_k_base,
                 section=FmhaStage.Head,
             )
+            if smem_q.cfg.stage_kv_by_head_dim:
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py]: can we support a general loop function like store O where tail mma will be handled automatically in the loop ? similar for other mmas

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py [hunk 7]
```diff
@@ -759,14 +786,15 @@ def mma_schedule(
...
 
-                    skv.wait()
-                    desc_v_base = skv.v_desc()
-                    to.pv_mma(
-                        desc_v_base=desc_v_base,
-                        section=FmhaStage.Loop,
-                        head_dim_stage_idx=1,
-                    )
-                    skv.release()
+                    if num_head_dim_stages_v == 2:
```
Comment (PerkzZheng) [inline flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py]: same, we should make this general with a loop instead of constant values for head dim 192 only.

**Review by PerkzZheng — COMMENTED** [review]
Finally, about the perf numbers in the MR description, you should only keep the latest prims_ts numbers + trtllm-gen, and please mention the unit (ms).

File: tests/attention/test_attention_ts_context_mla.py [hunk 8]
```diff
@@ -0,0 +1,199 @@
...
+# limitations under the License.
+
+"""Non-absorbed MLA: compact separate Q/K=192, V/O=128 on Blackwell."""
+
+import math
+
+import pytest
+import torch
+
+pytest.importorskip("cutlass", minversion="4.7.0")
```
[the remark this answers is not in the document]
Comment (PerkzZheng) [reply tests/attention/test_attention_ts_context_mla.py]: Fixed in b752577c. The MLA cases now live in tests/attention/test_attention_ts_context.py, whose gate matches the declared dependency minimum, 4.7.0a0. A collection check with prerelease version metadata succeeds; 4.6.0 is rejected. GPU validation uses the installed DSL 4.7.0.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 1 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:1889]: Previously the single-query D256 path used 128-wide K, V and O slices. Paired MLA keeps K/V transfers 128-wide, but two BF16 Q tiles plus a complete K/K/V cadence leave room for two 64-wide O buffers. FP8 operands or output can retain 128-wide O stores. The comment now explains this shared-memory distinction (b752577c).

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 2 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py:2510]: Added a comment identifying the logical Q/K and V head-dimension validation block before tile selection (b752577c).

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 3 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py]: Generalized the MMA extent to the next power of two with a 64-wide minimum: 64 maps to 64, 96 to 128, and 192 to 256. Global Q/K tensors remain compact (b752577c).

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py (hunk 4 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_kernel.py]: Ragged Q TMA now unconditionally uses cuda.TensorMapFloatOOBFill.none for row and head-dimension tails (b752577c).

File: tests/attention/test_attention_ts_context_mla.py
Comment (PerkzZheng) [reply tests/attention/test_attention_ts_context_mla.py:1]: Moved all 24 MLA cases into test_attention_ts_context.py, using its shared imports and per-test Blackwell guards. The separate test module was removed, and the README test command was updated (b752577c).

File: tests/attention/test_attention_ts_context.py (hunk 5 above)
Comment (PerkzZheng) [reply tests/attention/test_attention_ts_context.py:2646]: SM103 is already enabled by the capability tuple immediately above this branch and uses inline PTX when the native wrapper is unavailable. The SM107-only branch adds its separate native-wrapper/target requirement because DSL 4.7 targets sm_100f there. Added a comment clarifying this distinction (b752577c); GB300 validation exercises SM103.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py (hunk 6 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py]: Converted HEAD and both steady-state QK paths to loops over the configured K-stage count, with a shared slice helper and automatic logical-tail clipping in the common MMA implementation (b752577c). The admitted padded QK extents are at most 256, so the two 128-wide slices retain distinct scalar TS bindings. This preserves QK0/PV1/QK1/PV0 overlap without overwriting a descriptor still needed by the peer.

File: flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py (hunk 7 above)
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/fmha_tasks.py]: Replaced the explicit first/second K and V operations in the single-query steady state with loops over their configured stage counts (b752577c). This covers the two-K/one-V MLA path and the two-K/two-V D256 path through the same schedule code.

**Comment by PerkzZheng** [comment]
Addressed the review in b752577c:

- Generalized head padding and K/V stage loops while preserving separate descriptor bindings and QK/PV overlap.
- Made ragged-Q zero fill unconditional; clarified head validation, O staging, and SM103/SM107 reduction support.
- Consolidated all 24 MLA cases into the context suite, aligned the prerelease gate, and added docstrings to the cases and stage helpers.
- Updated the performance table to show only the latest prims_ts and trtllm-gen latencies, with ms in both column headers, as requested in the review summary.

Validation: pre-commit passed; all 180 context/mask tests passed; memcheck and synccheck each passed four cases with zero errors. All four BF16/FP8 8K/16K benchmark cases remain within 5% across three runs (worst paired gap 1.91%). Upstream GPU CI still awaits an authorized trigger.

File: flashinfer/attention/prims_ts/kernels/fmha_context/helpers.py [hunk 9]
```diff
@@ -94,3 +94,24 @@ def freeze_smem_descriptor(desc):
         write_only_types=[Int64],
         read_only_args=[desc],
     )
+
+
+_LDTM32_MAX_PTX = (
+    "tcgen05.ld.red.sync.aligned.32x32b.x32.f32.max {"
+    + ", ".join(f"{{$w{i}}}" for i in range(32))
+    + "}, {$w32}, [{$r0}];"
+)
```
[the remark this answers is not in the document]
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/helpers.py:103]: Checked at a01374e8: ptxas 13.3.73 accepts both `.32x32b.x32.f32.max` and `.32x32b.x32.max.f32` and produces byte-identical SM103 cubins (SHA-256 `7781df42aaa189b23564fbf2b773212b516db513a59c0d6288155333d69b1ff2`). The existing implementation also passes the full 291-case GB300 accuracy suite and both sanitizers. Keeping the validated spelling because the reported assembly failure was not reproduced.

File: flashinfer/attention/prims_ts/kernels/fmha_context/README.md [hunk 10]
```diff
@@ -57,8 +57,8 @@ alignment, value, aliasing, and lifetime obligation in the runtime contract.
 
 | Feature | Support |
 | --- | --- |
-| GPU | SM100a/B200 (qualified); SM103a/B300 (architecture-gated, not yet signoff-qualified) |
-| Head dimension | 128 or 256 |
+| GPU | Blackwell SM100 and SM103; see validation scope below |
```
[the remark this answers is not in the document]
Comment (PerkzZheng) [reply flashinfer/attention/prims_ts/kernels/fmha_context/README.md]: Fixed in a01374e8. The supported-GPU table now includes Rubin SM107, and the validation paragraph explicitly says this MLA extension has not been validated there. The existing SM100/SM103 validation details are preserved. Full pre-commit passed.

**Review by qsang-nv — APPROVED** [review]
